semantic-core/src/Analysis/Typecheck.hs

```diff
@@ -29,102 +27,84 @@ import qualified Data.IntSet as IntSet
+deriving instance (Show a, forall a . Show a => Show (f a))          => Show (Polytype f a)
+
+instance HFunctor Polytype
+instance RightModule Polytype where
+  PForAll b >>=* f = PForAll (b >>=* f)
 
-forAll :: Gensym -> Polytype -> Polytype
-forAll n body = PForAll (Analysis.Typecheck.bind n body)
 
-forAlls :: Foldable t => t Gensym -> Polytype -> Polytype
+forAll :: (Eq a, Carrier sig m, Member Polytype sig) => a -> m a -> m a
+forAll n body = send (PForAll (Data.Scope.bind1 n body))
+
+forAlls :: (Eq a, Carrier sig m, Member Polytype sig, Foldable t) => t a -> m a -> m a
 forAlls ns body = foldr forAll body ns
 
-generalize :: (Carrier sig m, Member Naming sig) => Monotype Meta -> m Polytype
-generalize ty = namespace "generalize" $ do
-  Gensym root _ <- Name.fresh
-  pure (forAlls (map (Gensym root) (IntSet.toList (mvs ty))) (fold root ty))
-  where fold root = \case
-          MUnit      -> PUnit
-          MBool      -> PBool
-          MString    -> PString
-          MMeta i    -> PFree (Gensym root i)
-          MFree n    -> PFree n
-          MArr a b   -> PArr (fold root a) (fold root b)
-          MRecord fs -> PRecord (fold root <$> fs)
-
--- | Bind occurrences of a 'Gensym' in a 'Polytype' term, producing a 'Scope' in which the 'Gensym' is bound.
-bind :: Gensym -> Polytype -> Scope
-bind name = Scope . substIn (\ i n -> if name == n then PBound i else PFree n) (const PBound)
-
--- | Substitute a 'Polytype' term for the free variable in a given 'Scope', producing a closed 'Polytype' term.
-instantiate :: Polytype -> Scope -> Polytype
-instantiate image (Scope body) = substIn (const PFree) (\ i j -> if i == j then image else PBound j) body
-
-substIn :: (Int -> Gensym -> Polytype)
-        -> (Int -> Int -> Polytype)
-        -> Polytype
-        -> Polytype
-substIn free bound = go 0
-  where go i (PFree name)           = free i name
-        go i (PBound j)             = bound i j
-        go i (PForAll (Scope body)) = PForAll (Scope (go (succ i) body))
-        go _ PUnit                  = PUnit
-        go _ PBool                  = PBool
-        go _ PString                = PString
-        go i (PArr a b)             = PArr (go i a) (go i b)
-        go i (PRecord fs)           = PRecord (go i <$> fs)
-
-
-typecheckingFlowInsensitive :: [File (Term Core.Core Name)] -> (Heap Name (Monotype Meta), [File (Either (Loc, String) Polytype)])
+generalize :: Term Monotype Meta -> Term (Polytype :+: Monotype) Void
+generalize ty = fromJust (closed (forAlls (IntSet.toList (mvs ty)) (hoistTerm R ty)))
+
+
+typecheckingFlowInsensitive :: [File (Term Core.Core Name)] -> (Heap Name (Term Monotype Meta), [File (Either (Loc, String) (Term (Polytype :+: Monotype) Void))])
 typecheckingFlowInsensitive
   = run
   . runFresh
   . runNaming
   . runHeap (Gen (Gensym (Nil :> "root") 0))
-  . (>>= traverse (traverse (traverse generalize)))
+  . fmap (fmap (fmap (fmap generalize)))
```

Comment:
…which is the result of generalizing (quantifying over all of the variables in)…